sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+        if (mode == "hit") {
+            entries.reserve(batch.hits.size());
+            for (const auto& hit : batch.hits) {
+                entries.push_back(static_cast<const IPositionTOF*>(&hit));
+            }
+        } else {
+            entries.reserve(batch.neutrons.size());
+            for (const auto& neutron : batch.neutrons) {
+                entries.push_back(static_cast<const IPositionTOF*>(&neutron));
+            }
+        }
+        
+        if (entries.empty()) {
+            spdlog::debug("Batch {} is empty", batch_index);
+            continue;
+        }
+
+        for (const auto& entry : entries) {
+            total_entries++;
+            double tof_ns = entry->iGetTOF_ns();
+
+            // Find the correct TOF bin
+            // NOTE: tof_bin_edges are in sec, and tof_ns are in nano secs
+            spdlog::debug("tof_ns: {}, tof_ns/1e9: {}", tof_ns, tof_ns/1e9);
+            auto it = std::lower_bound(tof_bin_edges.begin(), tof_bin_edges.end(), tof_ns/1e9);
+            if (it != tof_bin_edges.begin()) {
+                size_t bin_index = std::distance(tof_bin_edges.begin(), it) - 1;
+                
+                // Calculate the x and y indices in the 2D histogram
+                int x = static_cast<int>(entry->iGetX() * super_resolution);
+                int y = static_cast<int>(entry->iGetY() * super_resolution);
+                
+                // Ensure x and y are within bounds
+                if (x >= 0 && x < dim_x && y >= 0 && y < dim_y) {
+                    // Increment the count in the appropriate bin and position
+                    tof_images[bin_index][y][x]++;
+                    binned_entries++;
+                }
+            }
+        }
+    }
+
+    auto end = std::chrono::high_resolution_clock::now();
+    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+    spdlog::info("TOF image creation time: {} s", elapsed / 1e6);
+    spdlog::info("Total entries: {}, Binned entries: {}", total_entries, binned_entries);
+
+    return tof_images;
+}
+
+/**
+ * @brief Timed save TOF imaging to TIFF.
+ * 
+ * @param[in] out_tof_imaging
+ * @param[in] batches
+ * @param[in] tof_bin_edges
+ * @param[in] tof_filename_base
+ */
+void timedSaveTOFImagingToTIFF(
+    const std::string& out_tof_imaging,
+    const std::vector<std::vector<std::vector<unsigned int>>>& tof_images,
+    const std::vector<double>& tof_bin_edges,
+    const std::string& tof_filename_base)
+{
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // 1. Create output directory if it doesn't exist
+    if (!std::filesystem::exists(out_tof_imaging)) {
+        std::filesystem::create_directories(out_tof_imaging);
+        spdlog::info("Created output directory: {}", out_tof_imaging);
+    }
+
+    // 2. Initialize vector for spectral data
+    std::vector<unsigned long long> spectral_counts(tof_images.size(), 0);
```

Comment:
```suggestion
    std::vector<uint64_t> spectral_counts(tof_images.size(), 0);
```

Reply:
You are correct, we want consistent 64 bit to represent the count.